query/worker.go

```diff
@@ -0,0 +1,247 @@
+)
+
+// queryJob is the internal struct that wraps the Query to work on, in
+// addition to some information about the query.
+type queryJob struct {
+	index      uint64
+	timeout    time.Duration
+	encoding   wire.MessageEncoding
+	cancelChan <-chan struct{}
+	*Request
+}
+
+// queryJob should satisfy the Task interface in order to be sorted by the
+// workQueue.
+var _ Task = (*queryJob)(nil)
+
+// Index returns the queryJob's index within the work queue.
+//
+// NOTE: Part of the Task interface.
+func (q *queryJob) Index() uint64 {
+	return q.index
+}
+
+// jobResult is the final result of the worker's handling of the queryJob.
+type jobResult struct {
+	job  *queryJob
+	peer Peer
+	err  error
+}
+
+// worker is responsible for polling work from its work queue, and handing it
+// to the associated peer. It validates incoming responses with the current
+// query's response handler, and polls more work for the peer when it has
+// successfully received a response to the request.
+type worker struct {
+	peer Peer
+
+	// nextJob is a channel of queries to be distributed, where the worker
+	// will poll new work from.
+	nextJob chan *queryJob
+}
+
+// NewWorker creates a new worker assosiated with the given peer.
+func NewWorker(peer Peer) *worker {
+	return &worker{
+		peer:    peer,
+		nextJob: make(chan *queryJob),
+	}
+}
+
+// Run starts the worker. The worker will supply its with queries, and handle
+// responses from it.  Results for any query handled by this worker will be
+// delivered on the results channel. quit can be closed to immediately make the
+// worker exit.
+//
+// The method is blocking, and should be started in a goroutine. It will run
+// until the peer disconnects or the worker is told to quit.
+//
+// NOTE: Part of the Worker interface.
+func (w *worker) Run(results chan<- *jobResult,
+	quit <-chan struct{}) {
+
+	peer := w.peer
+
+	// Subscribe to messages from the peer.
+	msgChan, cancel := peer.SubscribeRecvMsg()
+	defer cancel()
+
+	for {
+		log.Tracef("Worker %v waiting for more work", peer.Addr())
+
+		var job *queryJob
+		select {
+		// Poll a new job from the nextJob channel.
+		case job = <-w.nextJob:
+			log.Tracef("Worker %v picked up job with index %v",
+				peer.Addr(), job.Index())
+
+		// Ignore any message received while not working on anything.
+		case msg := <-msgChan:
+			log.Tracef("Worker %v ignoring received msg %T "+
+				"since no job active", peer.Addr(), msg)
+			continue
+
+		// If the peer disconnected, we can exit immediately, as we
+		// weren't working on a query.
+		case <-peer.OnDisconnect():
+			log.Debugf("Peer %v for worker disconnected",
+				peer.Addr())
+			return
+
+		case <-quit:
+			return
+		}
+
+		select {
+		// There is no point in queueing the request if the job already
+		// is canceled, so we check this quickly.
+		case <-job.cancelChan:
+			log.Tracef("Worker %v found job with index %v "+
+				"already canceled", peer.Addr(), job.Index())
```

Comment:
Needs a `continue` after this?

Reply:
It should continue to send the response, added comment to make this clear: https://github.com/lightninglabs/neutrino/pull/179/commits/44a529ea3c830c15b5f5619709167102e9c31ea9